---
title: "An analysis of cutaneous leishmaniasis biopsies: `Sys.getenv('VERSION')`"
author: "Najib El-Sayed"
date: "`r Sys.Date()`"
output:
  html_document:
    code_download: true
    code_folding: show
    fig_caption: true
    fig_height: 7
    fig_width: 7
    highlight: zenburn
    keep_md: false
    mode: selfcontained
    number_sections: true
    self_contained: true
    theme: readable
    toc: true
    toc_float:
      collapsed: false
      smooth_scroll: false
---

<style type="text/css">
body, td {
font-size: 16px;
}
code.r{
font-size: 16px;
}
pre {
font-size: 16px
}
body .main-container {
max-width: 1600px;
}
</style>


```{r options, include=FALSE}
library(hpgltools)
library(glue)
knitr::opts_knit$set(
  progress = TRUE, verbose = TRUE, width = 90, echo = TRUE)
knitr::opts_chunk$set(
  error = TRUE, fig.width = 8, fig.height = 8, fig.retina = 2,
  fig.pos = "t", fig.align = "center", dpi = if (knitr::is_latex_output()) 72 else 300,
  out.width = "100%", dev = "png",
  dev.args = list("png" = list(type = "cairo-png")))
old_options <- options(
  digits = 4, stringsAsFactors = FALSE, knitr.duplicate.label = "allow")
ggplot2::theme_set(ggplot2::theme_bw(base_size = 12))
ver <- Sys.getenv("VERSION")
rundate <- format(Sys.Date(), format = "%Y%m%d")

rmd_file <- "01index.Rmd"
savefile <- gsub(pattern = "\\.Rmd", replace = "\\.rda\\.xz", x = rmd_file)
```

# Consolidation of a Molecular Signature of Healing in Cutaneous Leishmaniasis is Achieved During the First Ten Days of Treatment.

# Introduction

This document performs a series of diagnostics and differential
expression analyses which eventually helped inform the text of Lina
Fernanda Giraldo Parra's and Maria Adelaida Gomez's paper.  In it,
they sought to use biopsies over time from people who were observed to
cure/fail treatment of cutaneous leishmaniasis.  Some of these people
were treated with the antimonial 'glucantime', others with
miltefosine.  Biopsies were taken at the beginning, middle, and end of
treatment.  RNA was taken from the biopsies, made into libraries, and
sequenced.  The samples were mapped/counted against ensembl hg38
release 100 and the TritrypDB Leishmania panamensis MHOM/COL strain's
genome with hisat2 and salmon.  There were very few reads observed
from the parasite.  As a result (at the time of this writing: 202310),
those count tables are not included in the sample sheet (they are now,
along with the salmon quantitations against hg38_100), and are not
considered in this worksheet.

I collected the available metadata into the sheet 'all_samples.xlsx'.

# Sample sheet

```{r}
samplesheet <- "sample_sheets/all_samples.xlsx"
```

## Extracting extra sample annotations

The function 'gather_preprocessing_metadata()' scans the working
directory in which the various preprocessing tasks were performed,
trimming, fastqc/fastp, mapping/quantitation, variant searching, etc.
It is aware of my default output filenames for stdout/stderr/etc and
extracts from them various metrics potentially of interest.  I think
the sample sheet I copied into this container is the result of the
following block, so I am not going to evaluate it here.

```{r gather_preprocessing, eval=FALSE}
rna_spec <- make_rnaseq_spec()
modified <- gather_preprocessing_metadata(samplesheet,
                                          specification = rna_spec,
                                          species = "hg38_100")
```

# Annotation

We take the annotation data from ensembl's biomart instance.  The genome which
was used to map the data was hg38 revision 100.  My default when using biomart is
to load the data from 1 year before the current date.

```{r}
hs_annot <- load_biomart_annotations(year = "2019")
hs_annot

hs_annot <- hs_annot[["annotation"]]
hs_annot[["transcript"]] <- paste0(rownames(hs_annot), ".", hs_annot[["version"]])
rownames(hs_annot) <- make.names(hs_annot[["ensembl_gene_id"]], unique = TRUE)
tx_gene_map <- hs_annot[, c("transcript", "ensembl_gene_id")]
```

## Gene Ontology data

Downloading the Ontology annotation information has a significant
chance of failing because it sometimes takes longer than curl's
timeout of 300 seconds.  As a result, I may decide to include the
saved rda of ontology results in the container, or maybe improve my
ontology downloader so that it is no longer susceptible to timeouts...

Actually, I do not think I am using
clusterProfiler/gostats/goseq/topgo in this document, perhaps
therefore I should just exclude this block?

```{r}
hs_go <- load_biomart_go(overwrite = TRUE)
hs_go
hs_go <- hs_go[["go"]]
hs_length <- hs_annot[, c("ensembl_gene_id", "cds_length")]
colnames(hs_length) <- c("ID", "length")
```

# Create expressionsets

The primary datastructure used throughout this document is the
expressionSet or summarized Experiment.  This document uses the
former; but both are created by reading the sample sheet, extracting
the filenames of the count data from it, and combining the metadata,
annotation data, and counts.  My implementation of this process also
includes a few extra pieces of information.  For example: the colors
of various potential conditions in the data.

## Color choices

Many of the color choices are taken directly from another project
with CIDEIM which also looks at host responses.

```{r}
color_choices <- list(
  "cf_visit" = list(
    "cure_v1" = "#D95F0E",
    "failure_v1" = "#FEC44F",
    "cure_v2" = "#DD1C77",
    "failure_v2" = "#C994C7",
    "cure_v3" = "#3182BD",
    "failure_v3" = "#9ECAE1"),
  "drug_visit" = list(
    "antimoniate_v1" = "#badeef",
    "miltefosine_v1" = "#e4a6c1",
    "antimoniate_v2" = "#7bc6ea",
    "miltefosine_v2" = "#d26895",
    "antimoniate_v3" = "#1d91ca",
    "miltefosine_v3" = "#d52e75"),
  "species_visit" = list(
    "lvpanamensis_v1" = "#ffe798",
    "lvbraziliensis_v1" = "#b5b5b5",
    "lvpanamensis_v2" = "#FFC300",
    "lvbraziliensis_v2" = "#888888",
    "lvpanamensis_v3" = "#b58b00",
    "lvbraziliensis_v3" = "#525252"),
  "cf" = list(
    "cure" = "#998EC3",
    "failure" = "#F1A340"),
  "visit" = list(
    "v1" = "#33EE33",
    "v2" = "#11AA11",
    "v3" = "#134413"),
  "visitbi" = list(
    "v1" = "#BB0000",
    "vother" = "#0000BB"),
  "species" = list(
    "lvpanamensis" = "#FFC300",
    "lvbraziliensis" = "#525252"),
  "donor" = list(
    "d2008" = "#B78415",
    "d1029" = "#93752C",
    "d1036" = "#7E6EA2",
    "d1037" = "#B3499C",
    "d1031" = "#BD6332",
    "d2002" = "#7D8F31",
    "d2009" = "#8E7037",
    "d2010" = "#666666",
    "d1019" = "#1B9E77",
    "d2004" = "#E0A604",
    "d2001" =  "#CF3F76",
    "d2003" = "#A0A811"),
  "drug" = list(
    "antimoniate" = "#3182AA",
    "miltefosine" = "#C994AA"))
```

The set of color choices demonstrates the complexity of the
experimental design.  We are looking at combinations of time, outcome,
species, and drug.  Inherent in these is an unspecified donor effect.

## Initial dataset: Combine clinical outcome with visit

The initial datastructure will use a factor combining the clinical
outcome and visit as the experimental condition of interest.  We will
follow this up by splitting off the various factors of interest.

An aside: when testing the following block within the container, it
did not print out the filenames as it was reading them, usually it
does.  I assume this is not a problem since the data looks fine, but
this may warrant further exploration.

```{r}
wellcome_outtime <- create_expt(metadata = samplesheet, gene_info = hs_annot,
                                file_column = "hg38100hisatfile") %>%
  set_expt_batches(fact = "drug") %>%
  sanitize_expt_pData(spaces = TRUE)
wellcome_outtime

outcome_time <- paste0(pData(wellcome_outtime)[["clinicaloutcome"]], "_v",
                       pData(wellcome_outtime)[["visitnumber"]])
wellcome_outtime <- set_expt_conditions(wellcome_outtime, fact = outcome_time,
                                        colors = color_choices[["cf_visit"]])
pData(wellcome_outtime)[["visitnumber"]] <- paste0("v", pData(wellcome_outtime)[["visitnumber"]])

pData(wellcome_outtime)[["drug_visit"]] <- paste0(pData(wellcome_outtime)[["drug"]], "_",
                                                  pData(wellcome_outtime)[["visitnumber"]])
pData(wellcome_outtime)[["species_visit"]] <- paste0(pData(wellcome_outtime)[["infectingspecies"]], "_",
                                                     pData(wellcome_outtime)[["visitnumber"]])

wellcome_outtime <- sanitize_expt_pData(
  wellcome_outtime, spaces = TRUE,
  columns = c("clinicaloutcome", "visitnumber", "infectingspecies", "drug"))
```

## Check samples

First, let us get a quick view of the samples in their native state.
There are a few samples which are candidates for removal due to lower
coverage.

Random aside: I have some nifty code which tries to autodetect when to
color the text in the colored bars white or black.  The purple/pink
color tricks that code into thinking it is dark enough to be colored white.

```{r}
plot_legend(wellcome_outtime)
plot_libsize(wellcome_outtime)
plot_nonzero(wellcome_outtime, plot_labels = "repel")

wellcome_filtered <- subset_expt(wellcome_outtime, nonzero = 15000)
```

Looking at the nonzero plot, it seems that 15k genes is a reasonable
cutoff, which would remove 2 samples.  So, for the moment I will split
the data up into two sets, pre/post removal.

## Upload CPM values

Lina asked for a copy of the data as CPM.

```{r}
wellcome_cpm <- normalize_expt(wellcome_filtered, filter = TRUE, convert = "cpm")
wellcome_cpm_mtrx <- exprs(wellcome_cpm)
wellcome_cpm_write <- write.csv(x = wellcome_cpm_mtrx, file = "excel/wellcome_fc.csv")

wellcome_scpm <- normalize_expt(wellcome_filtered, filter = TRUE, convert = "cpm", batch = "svaseq")
wellcome_scpm_mtrx <- exprs(wellcome_scpm)
wellcome_scpm_write <- write.csv(x = wellcome_cpm_mtrx, file = "excel/wellcome_fc_cpm_sva.csv")
```

## Check samples

We have in place some initial datastructures; let us mix and match
them to get a sense of the data.  There are a few factors in the
experiment that are likely of primary interest: the donor, time, drug
used, parasite species/strain, and clinical outcome.  There are too
many factors for the number of samples to do a full rank model, so I
am going to make copies of the expressionset and model each factor
separately with the help of sva.

## Only donor

Start with the donor.  There are 12 people, most of whom provided
three samples. Two of them (d2002 and d2009) had samples which got
excluded and so have only 2.  I am separating the un/filtered datasets
so that we may play with both if we want.

```{r}
wellcome_donor <- set_expt_conditions(wellcome_outtime, fact = "donor",
                                      colors = color_choices)

wellcome_filtdonor <- set_expt_conditions(wellcome_filtered, fact = "donor",
                                          colors = color_choices)
```

## Only visit

In this case, the condition will only be visit number.  We have 12
samples for each visit, except v1/v3 which are missing 1 each.

```{r}
wellcome_time <- set_expt_conditions(wellcome_outtime, fact = "visitnumber",
                                     colors = color_choices[["visit"]])

wellcome_filttime <- set_expt_conditions(wellcome_filtered, fact = "visitnumber",
                                         colors = color_choices[["visit"]])
```

## Only Cure/Fail

Conversely, we can split by clinical outcome.  The unfiltered dataset
has 15 cures and 21 failures; both filtered samples were failure.

```{r}
wellcome_outcome <- set_expt_conditions(wellcome_outtime, fact = "clinicaloutcome",
                                        colors = color_choices[["cf"]]) %>%
  set_expt_batches(fact = "visitnumber")

wellcome_outcome_filt <- set_expt_conditions(wellcome_filtered, fact = "clinicaloutcome",
                                             colors = color_choices[["cf"]]) %>%
  set_expt_batches(fact = "visitnumber")
```

## Only drug

With respect to drug treatment, we started with 18 of each.  The lost
samples were both miltefosine.

```{r}
wellcome_drug <- set_expt_conditions(wellcome_outtime, fact = "drug",
                                     colors = color_choices[["drug"]]) %>%
  set_expt_batches(fact = "visitnumber")

wellcome_drugfilt <- set_expt_conditions(wellcome_filtered, fact = "drug",
                                         colors = color_choices[["drug"]]) %>%
  set_expt_batches(fact = "visitnumber")
```

## Only parasite

We started with 12 braziliensis and 24 panamensis; one of each was lost.

```{r}
wellcome_parasite <- set_expt_conditions(wellcome_outtime, fact = "infectingspecies",
                                         colors = color_choices[["species"]]) %>%
  set_expt_batches(fact = "visitnumber")

wellcome_parafilt <- set_expt_conditions(wellcome_filtered, fact = "infectingspecies",
                                         colors = color_choices[["species"]]) %>%
  set_expt_batches(fact = "visitnumber")
```

## Visualize the metadata

Everything I wrote above is visible in the following sankey plot.  It
is particularly noteworthy that we have no cure braziliensis for
miltefosine treatment and only 1 for each of the antominial
timepoints.

```{r}
drug_visit_species_cf <- plot_meta_sankey(
  wellcome_outtime, factors = c("drug", "visitnumber", "clinicaloutcome", "infectingspecies"),
  color_choices = color_choices)
drug_visit_species_cf
```

# Visualize samples

Given the large number of variables in this data, lets start by
getting a feeling for the amount of variance in each.  Given that we
don't have a full rank with respect to clinical outcome, I assume that
trying variance partition with the 4 primary factors will fail, but
let us see...

```{r}
all_varpart_donor <- simple_varpart(
  wellcome_filtered,
  factors = c("donor", "visitnumber"))
all_varpart_donor

all_varpart_dvic <- simple_varpart(
  wellcome_filtered,
  factors = c("drug", "visitnumber", "infectingspecies", "clinicaloutcome"))
all_varpart_dvic
```

To my eyes, it appears that donor is dominant factor,
followed by: visit, species, drug, and outcome.  I think this
observation may have an effect on the current state of the
manuscript - which if I recall properly states that donor is not a
significant factor in the data.

## Examine top-n genes with respect to variance of some factors

Let us ask if there are categories associated with the genes
associated with each of these categories and see if they 'make sense'.

```{r}
top_300_donor_genes_idx <- order(all_varpart_donor[["fitted_df"]][["donor"]])
top_300_donor_genes <- tail(all_varpart_donor[["fitted_df"]][top_300_donor_genes_idx, ], n = 300)
summary(top_300_donor_genes)
donor_genes_gp <- simple_gprofiler(rownames(top_300_donor_genes))
donor_genes_gp
donor_genes_gp$pvalue_plots$REAC
```

So, the top-300 genes with respect to putative donor effect, when
passed to gprofiler's over representation analyses, look like they
have lots of variance related to categories that are explicitly
important to the general questions we are asking.  I think that counts
as: Yay!

Now let us look at the other factors in the data and see if anything
noteworthy pops out.  Given that the dvic datastructure comprises the
other factors of interest, we will need to reorder the results with
respect to each factor separately.

### Visit

Visit also is dominated by variance which is explicitly what one would expect:
wound healing.

```{r}
top_300_visit_genes_idx <- order(all_varpart_dvic[["fitted_df"]][["visitnumber"]])
top_300_visit_genes <- tail(all_varpart_dvic[["fitted_df"]][top_300_visit_genes_idx, ], n = 300)
summary(top_300_visit_genes)
visit_genes_gp <- simple_gprofiler(rownames(top_300_visit_genes))
visit_genes_gp
```

### Drug

I do not have any idea what one should expect vis a vis the two drugs.
Looking at the following result, I guess one should not be surprised
given that I am unaware of m/any experiments which explicitly compare
antimonial treatments with respect to transcriptional profile.

```{r}
top_300_drug_genes_idx <- order(all_varpart_dvic[["fitted_df"]][["drug"]])
top_300_drug_genes <- tail(all_varpart_dvic[["fitted_df"]][top_300_drug_genes_idx, ], n = 300)
summary(top_300_drug_genes)
drug_genes_gp <- simple_gprofiler(rownames(top_300_drug_genes))
drug_genes_gp
drug_genes_gp$pvalue_plots$MF
```

### Infecting species

The violin plot above suggests there is very limited variance with
respect to the two species.  In addition, there are significantly
fewer braziliensis samples which failed treatment (for miltefosine in
particular), so I presume variancePartition is going to have trouble
extracting genes which are reliable in this context.  In addition, I
would assume that the human transcriptome has pretty similar responses
to the two parasites, exacerbating this confounder.

```{r}
top_300_species_genes_idx <- order(all_varpart_dvic[["fitted_df"]][["infectingspecies"]])
top_300_species_genes <- tail(all_varpart_dvic[["fitted_df"]][top_300_species_genes_idx, ], n = 300)
summary(top_300_species_genes)
species_genes_gp <- simple_gprofiler(rownames(top_300_species_genes))
species_genes_gp
```

### Clinical outcome

In my mind, clinical outcome is the most generally interesting query.
It is also one with limited information in the data, _but_ lots of
other studies have been performed in this realm, so it is more likely
to have reliable overrepresentation results even with limited
information.  Thus, when I look at the bar/dot plot the categories look
interesting.

```{r}
top_300_cf_genes_idx <- order(all_varpart_dvic[["fitted_df"]][["clinicaloutcome"]])
top_300_cf_genes <- tail(all_varpart_dvic[["fitted_df"]][top_300_cf_genes_idx, ], n = 300)
summary(top_300_cf_genes)
cf_genes_gp <- simple_gprofiler(rownames(top_300_cf_genes))
cf_genes_gp
enrichplot::dotplot(cf_genes_gp$GO_enrich)
```

# The samples' distribution is similar without 2 samples

Here are the PCA results before/after removing the two samples that I
called shenanigans on.  They are quite similar.

```{r}
wellcome_outtime_norm <- normalize_expt(wellcome_outtime, filter = TRUE,
                                        convert = "cpm", transform = "log2", norm = "quant")
plot_pca(wellcome_outtime_norm)

wellcome_outfilt_norm <- normalize_expt(wellcome_filtered, filter = TRUE,
                                        convert = "cpm", transform = "log2", norm = "quant")
plot_pca(wellcome_outfilt_norm)
```

# GSVA

From my perspective, it seems that the most interesting GSVA
comparison is to look for scores changing between the cure and fail
samples.  The following block therefore passes the raw expression data
and performs the gene set variance analysis of it against the mSigDB
C2 set of categories by default.  We can pretty easily change the
mSigDB release/category set; except I would need download the dataset
to the container to use a newer revision.

```{r}
wt_gsva <- simple_gsva(wellcome_outcome_filt, signature_category = "c7")
wt_gsva

wt_gsva_sig <- get_sig_gsva_categories(wt_gsva, excel = "excel/wellcome_filter_outcome_gsva.xlsx")
wt_gsva_sig
```

Hmm, did I get confused, is C7 what I thought it was?  Yeah, it is the
immunologic signature sets.

# Visualize Various PCA

Now that we have an initial sense of how the samples relate to each
other, let us poke a bit further.  In each of the following, I am
likely to do one plot before and one after using sva.

## By visit

```{r}
wellcome_time_norm <- normalize_expt(wellcome_filttime, norm = "quant", convert = "cpm",
                                     filter = TRUE, transform = "log2")
plot_pca(wellcome_time_norm)

wellcome_time_nb <- normalize_expt(wellcome_filttime, convert = "cpm",
                                   batch = "svaseq", filter = TRUE, transform = "log2")
plot_pca(wellcome_time_nb)
```

## Cure/Fail

```{r}
wellcome_outcome_norm <- normalize_expt(wellcome_outcome_filt, norm = "quant", convert = "cpm",
                                        filter = TRUE, transform = "log2")
plot_pca(wellcome_outcome_norm, plot_labels = FALSE)

wellcome_outcome_nb <- normalize_expt(wellcome_outcome_filt, convert = "cpm",
                                      batch = "svaseq", filter = TRUE, transform = "log2")
plot_pca(wellcome_outcome_nb)
```

Wow, there really isn't much C/F variance, is there.

## Drug

```{r}
wellcome_drug_norm <- normalize_expt(wellcome_drugfilt, norm = "quant", convert = "cpm",
                                     filter = TRUE, transform = "log2")
plot_pca(wellcome_drug_norm)

wellcome_drug_nb <- normalize_expt(wellcome_drugfilt, convert = "cpm",
                                   batch = "svaseq", filter = TRUE, transform = "log2")
plot_pca(wellcome_drug_nb)
```

### Separate drug treatment by time

Conversely, it is likely that we want to separate and/or combine the
various factors with the visit.

#### Single factor all samples

```{r}
drugtime <- set_expt_conditions(wellcome_outtime, fact = "drug_visit",
                                colors = color_choices[["drug_visit"]]) %>%
  set_expt_batches(fact = "clinicaloutcome")
drugtime_norm <- normalize_expt(drugtime, norm = "quant", convert = "cpm",
                                transform = "log2", filter = TRUE)
plot_pca(drugtime_norm)

drugtime_nb <- normalize_expt(drugtime, batch = "svaseq", convert = "cpm",
                              transform = "log2", filter = TRUE)
plot_pca(drugtime_nb)
```

#### Visit 1

One interpretation of some queries from reviewers would be to replot
the various PCAs with the visits separated from each other.  I am not
sure if that is the correct interpretation, but here it is
nonetheless.

```{r}
drug_v1 <- subset_expt(drugtime, subset = "visitnumber=='v1'") %>%
  set_expt_batches(fact = "clinicaloutcome")
drug_v1_norm <- normalize_expt(drug_v1, norm = "quant", convert = "cpm",
                               filter = TRUE, transform = "log2")
plot_pca(drug_v1_norm)
drug_v1_nb <- normalize_expt(drug_v1, batch = "svaseq", convert = "cpm",
                               filter = TRUE, transform = "log2")
plot_pca(drug_v1_nb)
```

#### Visit 2

```{r}
drug_v2 <- subset_expt(drugtime, subset = "visitnumber=='v2'") %>%
  set_expt_batches(fact = "clinicaloutcome")
drug_v2_norm <- normalize_expt(drug_v2, norm = "quant", convert = "cpm",
                               filter = TRUE, transform = "log2")
plot_pca(drug_v2_norm)
drug_v2_nb <- normalize_expt(drug_v2, batch = "svaseq", convert = "cpm",
                               filter = TRUE, transform = "log2")
plot_pca(drug_v2_nb)
```

#### Visit 3

```{r}
drug_v3 <- subset_expt(drugtime, subset = "visitnumber=='v3'") %>%
  set_expt_batches(fact = "clinicaloutcome")
drug_v3_norm <- normalize_expt(drug_v3, norm = "quant", convert = "cpm",
                               filter = TRUE, transform = "log2")
plot_pca(drug_v3_norm)
drug_v3_nb <- normalize_expt(drug_v3, batch = "svaseq", convert = "cpm",
                               filter = TRUE, transform = "log2")
plot_pca(drug_v3_nb)
```

## Outcome and time

Now let us combine C/F, time and repeat the process.

```{r}
wellcome_outtime_norm <- normalize_expt(wellcome_filtered, norm = "quant", convert = "cpm",
                                        filter = TRUE, transform = "log2")
plot_pca(wellcome_outtime_norm)

wellcome_outtime_nb <- normalize_expt(wellcome_filtered, convert = "cpm",
                                      batch = "svaseq", filter = TRUE, transform = "log2")
plot_pca(wellcome_outtime_nb)
```

#### Visit 1

```{r}
outtime_v1 <- subset_expt(wellcome_outtime, subset = "visitnumber=='v1'") %>%
  set_expt_batches(fact = "drug")
outtime_v1_norm <- normalize_expt(outtime_v1, norm = "quant", convert = "cpm",
                               filter = TRUE, transform = "log2")
plot_pca(outtime_v1_norm)

outtime_v1_nb <- normalize_expt(outtime_v1, batch = "svaseq", convert = "cpm",
                               filter = TRUE, transform = "log2")
plot_pca(outtime_v1_nb)
```

#### Visit 2

```{r}
outtime_v2 <- subset_expt(wellcome_outtime, subset = "visitnumber=='v2'") %>%
  set_expt_batches(fact = "drug")
outtime_v2_norm <- normalize_expt(outtime_v2, norm = "quant", convert = "cpm",
                               filter = TRUE, transform = "log2")
plot_pca(outtime_v2_norm)

outtime_v2_nb <- normalize_expt(outtime_v2, batch = "svaseq", convert = "cpm",
                               filter = TRUE, transform = "log2")
plot_pca(outtime_v2_nb)
```

#### Visit 3

```{r}
outtime_v3 <- subset_expt(wellcome_outtime, subset = "visitnumber=='v3'") %>%
  set_expt_batches(fact = "drug")
outtime_v3_norm <- normalize_expt(outtime_v3, norm = "quant", convert = "cpm",
                               filter = TRUE, transform = "log2")
plot_pca(outtime_v3_norm)

outtime_v3_nb <- normalize_expt(outtime_v3, batch = "svaseq", convert = "cpm",
                               filter = TRUE, transform = "log2")
plot_pca(outtime_v3_nb)
```

## Parasite Species

### Separate species by time

#### Single factor all samples

```{r}
speciestime <- set_expt_conditions(wellcome_outtime, fact = "species_visit",
                                   colors = color_choices[["species_visit"]]) %>%
  set_expt_batches(fact = "clinicaloutcome")

speciestime_norm <- normalize_expt(speciestime, norm = "quant", convert = "cpm",
                                   transform = "log2", filter = TRUE)
plot_pca(speciestime_norm)
speciestime_nb <- normalize_expt(speciestime, batch = "svaseq", convert = "cpm",
                                   transform = "log2", filter = TRUE)
plot_pca(speciestime_nb)
```

#### Visit 1

```{r}
species_v1 <- subset_expt(speciestime, subset = "visitnumber=='v1'") %>%
  set_expt_batches(fact = "clinicaloutcome")
species_v1_norm <- normalize_expt(species_v1, norm = "quant", convert = "cpm",
                               filter = TRUE, transform = "log2")
plot_pca(species_v1_norm)
species_v1_nb <- normalize_expt(species_v1, batch = "svaseq", convert = "cpm",
                               filter = TRUE, transform = "log2")
plot_pca(species_v1_nb)
```

#### Visit 2

```{r}
species_v2 <- subset_expt(speciestime, subset = "visitnumber=='v2'") %>%
  set_expt_batches(fact = "clinicaloutcome")
species_v2_norm <- normalize_expt(species_v2, norm = "quant", convert = "cpm",
                                  filter = TRUE, transform = "log2")
plot_pca(species_v2_norm)
species_v2_nb <- normalize_expt(species_v2, batch = "svaseq", convert = "cpm",
                               filter = TRUE, transform = "log2")
plot_pca(species_v2_nb)
```

#### Visit 3

```{r}
species_v3 <- subset_expt(speciestime, subset = "visitnumber=='v3'") %>%
  set_expt_batches(fact = "clinicaloutcome")
species_v3_norm <- normalize_expt(species_v3, norm = "quant", convert = "cpm",
                               filter = TRUE, transform = "log2")
plot_pca(species_v3_norm)
species_v3_nb <- normalize_expt(species_v3, batch = "svaseq", convert = "cpm",
                               filter = TRUE, transform = "log2")
plot_pca(species_v3_nb)
```

#### Back to everything

```{r}
wellcome_parasite_norm <- normalize_expt(wellcome_parafilt, norm = "quant", convert = "cpm",
                                         filter = TRUE, transform = "log2")
plot_pca(wellcome_parasite_norm)

wellcome_parasite_nb <- normalize_expt(wellcome_parafilt, norm = "quant", convert = "cpm",
                                       batch = "svaseq", filter = TRUE, transform = "log2")
plot_pca(wellcome_parasite_nb)
```

Ok, so that is a big pile of pictures, now let us perform some DE
analyses and see what pops out.

# DE analyses

Given the above variance partition and PCA plots, we can surmise that
some metadata factors are much more likely to give interesting results
than others (Drug far more than C/F, for example).  With that in mind,
let us perform the various possible DE analyses with each
factor and see what comes out.  Each of these runs of all pairwise
will be performed once with SVA estimates in the model, and once with
the drug treatment as the batch factor.

## Outcome and time

## The contrasts of interest

The following list names each contrast and the two element vector
following provides each numerator and denominator.

```{r}
outtime_keepers <- list(
  "curev1v2" = c("curev2", "curev1"),
  "curev1v3" = c("curev3", "curev1"),
  "curev2v3" = c("curev3", "curev2"),
  "failv1v2" = c("failurev2", "failurev1"),
  "failv1v3" = c("failurev3", "failurev1"),
  "failv2v3" = c("failurev3", "failurev2"),
  "cfv1" = c("failurev1", "curev1"),
  "cfv2" = c("failurev2", "curev2"),
  "cfv3" = c("failurev3", "curev3"))
```

Start with the combined factor of {outcome}_{visit}.  Given the PCA, I
expect to see a little bit of information with batch in the model,
oddly less information with SVA.

### SVA

One of the review comments seemed to me to suggest that using the
variancePartition dream method might be useful.  I am not certain if
that is currently enabled.  I definitely got it working, though.

```{r}
outtime_sva_de <- all_pairwise(wellcome_filtered, model_batch = "svaseq", filter = TRUE)
outtime_sva_de
outtime_sva_table <- combine_de_tables(
  outtime_sva_de,
  keepers = outtime_keepers,
  excel = glue("excel/wellcome_outtime_table_sva-v{ver}.xlsx"))
outtime_sva_table
outtime_sva_sig <- extract_significant_genes(
  outtime_sva_table,
  excel = glue("excel/wellcome_outtime_sig_sva-v{ver}.xlsx"))
outtime_sva_sig
```

### Batch in model

```{r}
outtime_batch_de <- all_pairwise(wellcome_filtered, model_batch = TRUE, filter = TRUE)
outtime_batch_de
outtime_batch_table <- combine_de_tables(
  outtime_batch_de,
  keepers = outtime_keepers,
  excel = glue("excel/wellcome_outtime_table_batch-v{ver}.xlsx"))
outtime_batch_table
outtime_batch_sig <- extract_significant_genes(
  outtime_batch_table,
  excel = glue("excel/wellcome_outtime_sig_batch-v{ver}.xlsx"))
outtime_batch_sig
```

# PROPER

There was a request some time ago to provide a power analysis.  I
arbitrarily chose this dataset in order to invoke PROPER and provide
the resulting plots...

```{r}
outtime_batch_proper <- simple_proper(outtime_batch_table)
outtime_batch_proper[["curev2_vs_curev1"]][["power_table"]]
outtime_batch_proper[["curev2_vs_curev1"]][["power_plot"]]
outtime_batch_proper[["curev2_vs_curev1"]][["powertd_plot"]]
outtime_batch_proper[["curev2_vs_curev1"]][["powerfd_plot"]]
outtime_batch_proper[["curev2_vs_curev1"]][["fdcost_plot"]]
outtime_batch_proper[["curev2_vs_curev1"]][["powerhist_plot"]]
outtime_batch_proper[["curev2_vs_curev1"]][["poweralpha_plot"]]
```

## Only time

Now let us compare the time points, with and without SVA.

## Time contrasts

```{r}
time_keepers <- list(
  "v1v2" = c("v2", "v1"),
  "v1v3" = c("v3", "v1"),
  "v2v3" = c("v3", "v2"))
```

### SVA

```{r}
time_sva_de <- all_pairwise(wellcome_filttime, model_batch = "svaseq", filter = TRUE)
time_sva_de
time_sva_table <- combine_de_tables(
  time_sva_de,
  keepers = time_keepers,
  excel = glue("excel/wellcome_time_table_sva-v{ver}.xlsx"))
time_sva_table
time_sva_sig <- extract_significant_genes(
  time_sva_table,
  excel = glue("excel/wellcome_time_sig_sva-v{ver}.xlsx"))
time_sva_sig
```

### Batch

```{r}
time_batch_de <- all_pairwise(wellcome_filttime, model_batch = "batchseq", filter = TRUE)
time_batch_de
time_batch_table <- combine_de_tables(
  time_batch_de,
  keepers = time_keepers,
  excel = glue("excel/wellcome_time_table_batch-v{ver}.xlsx"))
time_batch_table
time_batch_sig <- extract_significant_genes(
  time_batch_table,
  excel = glue("excel/wellcome_time_batch_sva-v{ver}.xlsx"))
time_batch_sig
```

## Strain

## Contrasts of interest

I neglected to set a contrast name/value for this.

```{r}
## Oh, I just let the computer choose the numerator/denominator on its own.
```

The infecting strain I think should prove one of the more interesting comparisons.

```{r}
parasite_sva_de <- all_pairwise(wellcome_parafilt, model_batch = "svaseq", filter = TRUE)
parasite_sva_de
parasite_sva_table <- combine_de_tables(
  parasite_sva_de,
  excel = glue("excel/wellcome_parasite_table_sva-v{ver}.xlsx"))
parasite_sva_table
parasite_sva_sig <- extract_significant_genes(
  parasite_sva_table,
  excel = glue("excel/wellcome_parasite_sig_sva-v{ver}.xlsx"))
parasite_sva_sig

parasite_batch_de <- all_pairwise(wellcome_parafilt, model_batch = "batchseq", filter = TRUE)
parasite_batch_de
parasite_batch_table <- combine_de_tables(
  parasite_batch_de,
  excel = glue("excel/wellcome_parasite_table_batch-v{ver}.xlsx"))
parasite_batch_table
parasite_batch_sig <- extract_significant_genes(
  parasite_batch_table,
  excel = glue("excel/wellcome_parasite_sig_sva-v{ver}.xlsx"))
parasite_batch_sig
```

## Outcome

```{r}
outcome_sva_de <- all_pairwise(wellcome_outcome_filt, model_batch = "svaseq", filter = TRUE)
outcome_sva_de
outcome_sva_table <- combine_de_tables(
  outcome_sva_de,
  excel = glue("excel/wellcome_outcome_table_sva-v{ver}.xlsx"))
outcome_sva_table
outcome_sva_sig <- extract_significant_genes(
  outcome_sva_table,
  excel = glue("excel/wellcome_outcome_sig_sva-v{ver}.xlsx"))
outcome_sva_sig

outcome_batch_de <- all_pairwise(wellcome_outcome_filt, model_batch = "batchseq", filter = TRUE)
outcome_batch_de
outcome_batch_table <- combine_de_tables(
  outcome_batch_de,
  excel = glue("excel/wellcome_outcome_table_batch-v{ver}.xlsx"))
outcome_batch_table
outcome_batch_sig <- extract_significant_genes(
  outcome_batch_table,
  excel = glue("excel/wellcome_outcome_sig_sva-v{ver}.xlsx"))
outcome_batch_sig
```

# Additional GSVA

Another round of GSVA, I am not sure why this is here?

```{r}
wellcome_gsva_c2 <- simple_gsva(wellcome_filtered, signature_category = "c2")
wellcome_gsva_c2_sig <- get_sig_gsva_categories(
  wellcome_gsva_c2,
  excel = "excel/wellcome_gsva_c2.xlsx")

wellcome_gsva_c7 <- simple_gsva(wellcome_filtered, signature_category = "c7")
wellcome_gsva_c7_sig <- get_sig_gsva_categories(
  wellcome_gsva_c7,
  excel = "excel/wellcome_gsva_c7.xlsx")
```

```{r}
outtime_gprofiler <- all_gprofiler(outtime_sva_sig)
outtime_gprofiler$curev1v2_up$pvalue_plots$bpp_plot_over
outtime_gprofiler$failv1v2_up$pvalue_plots$kegg_plot_over

outtime_gprofiler$cfv1_up$pvalue_plots$reactome_plot_over
outtime_gprofiler$cfv1_down$pvalue_plots$reactome_plot_over
```
